vllm/model_executor/custom_op.py

```diff
@@ -1,52 +1,106 @@
 
+    def forward_native(self, *args, **kwargs):
+        """PyTorch-native implementation of the forward method.
+
+        This method is expected to invoke the forward_static method
+        with the input arguments and the attributes of `self`.
+        By default, we assume that forward_static does not need any
+        attributes of `self`.
+        """
+        return self.forward_static(*args, **kwargs)
+
+    def forward_compile(self, *args, **kwargs):
+        """Runs a torch-compiled version of forward_native.
+
+        This method hooks into the forward_native method and compiles
+        `forward_static` using `torch.compile` if it has not been compiled yet.
+        """
+        if not self._is_compiled and not envs.VLLM_TEST_DYNAMO_GRAPH_CAPTURE:
+            self.forward_static = torch.compile(  # type: ignore
+                self.forward_static)
+            self._is_compiled = True
+        return self.forward_native(*args, **kwargs)
+
     def forward_cuda(self, *args, **kwargs):
-        raise NotImplementedError
+        """Forward method for NVIDIA GPUs.
+
+        By default, we use torch.compile to optimize the op. However, we can
+        override this method to use a custom CUDA implementation if needed.
+        """
+        return self.forward_compile(*args, **kwargs)
 
     def forward_hip(self, *args, **kwargs):
-        # By default, we assume that HIP ops are compatible with CUDA ops.
+        """Forward method for AMD GPUs.
+
+        By default, this method is the same as forward_cuda. However, we can
+        override this method to use a custom HIP implementation if needed.
+        """
         return self.forward_cuda(*args, **kwargs)
 
     def forward_xpu(self, *args, **kwargs):
-        raise NotImplementedError
+        """Forward method for Intel XPUs.
+
+        By default, we use torch.compile to optimize the op. However, we can
+        override this method to use a custom XPU implementation if needed.
+        """
+        return self.forward_compile(*args, **kwargs)
 
     def forward_cpu(self, *args, **kwargs):
-        # By default, we assume that CPU ops are compatible with CUDA ops.
-        return self.forward_cuda(*args, **kwargs)
+        """Forward method for CPUs.
 
-    def forward_tpu(self, *args, **kwargs):
-        # By default, we assume that TPU ops are compatible with the
-        # PyTorch-native implementation.
-        # NOTE(woosuk): This is a placeholder for future extensions.
+        By default, we use torch.compile to optimize the op. However, we can
+        override this method to use a custom CPU implementation if needed.
+        """
+        if not self._is_compiled and not envs.VLLM_TEST_DYNAMO_GRAPH_CAPTURE:
+            self.forward_static = torch.compile(  # type: ignore
+                self.forward_static,
+                options={
+                    "fx_graph_cache": True,
+                    "cpp_wrapper": True,
+                    "dce": True,
```

Comment:
I'd be careful with DCE and inplace ops.  I've run into trouble with this.

Reply:
Thanks for the heads up. Removed the flag.